### ISSUE TITLE
Typings for AssignmentProperty could use improvement

### ISSUE THREAD

**Issue by axefrog** [cherow/cherow#280]
Automatic type inference is one the key benefits provided by type definitions, but currently the `AssignmentPattern` definition, while technically correct, isn't very useful with respect to what it should represent when narrowing types in the surrounding logic.

```ts
export interface AssignmentProperty extends _Node<'Property'> {
  key: Expression;
  value: PatternNoRest;
  computed: boolean;
  kind: 'init';
  method: false;
  shorthand: boolean;
}
```

- `key` is only `Expression` if `computed` is `true`
- If `computed` is `false`, or `shorthand` is `true`, then `key` is always `Identifier`
- If `computed` is `true`, `value` is always `ObjectPattern`, `ArrayPattern` or `AssignmentPattern`
- If `shorthand` is `true`, `value` is always `Identifier`

If the above could be worked into the type definitions, any logic that checks the above properties will automatically narrow the associated types, making it easier to rely on inference and avoid casting.

**Comment by KFlash** [comment]
This make sense,but isn't this this how ESTree already do it?

**Comment by axefrog** [comment]
You are right. I misinterpreted `AssignmentProperty` as an internal typing that narrowed an ESTree-specified type. Do you think there is any potential to improve certain classes of typings (such as the one described by this issue) in a way that retains ESTree compliance but also improved the developer experience when using cherow?

**Comment by KFlash** [comment]
@axefrog [this](parseAssignmentProperty) is how `AssignmentProperty` have been implemented in 2.0
